libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,560 @@
+			Stack stack = _functionReturn.info->returnVariables | ranges::views::transform([](auto const& _varSlot){
+				return StackSlot{_varSlot};
+			}) | ranges::to<Stack>;
+			stack.emplace_back(FunctionReturnLabelSlot{});
+			return stack;
+		},
+		[&](CFG::BasicBlock::Terminated const&) -> std::optional<Stack>
+		{
+			return Stack{};
+		},
+	}, _block.exit);
+}
+
+void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const& _entry)
+{
+	list<CFG::BasicBlock const*> toVisit{&_entry};
+	set<CFG::BasicBlock const*> visited;
+
+	// TODO: check whether visiting only a subset of these in the outer iteration below is enough.
+	list<pair<CFG::BasicBlock const*, CFG::BasicBlock const*>> backwardsJumps = collectBackwardsJumps(_entry);
+
+	while (!toVisit.empty())
+	{
+		// First calculate stack layouts without walking backwards jumps, i.e. assuming the current preliminary
+		// entry layout of the backwards jump target as the initial exit layout of the backwards-jumping block.
+		while (!toVisit.empty())
+		{
+			CFG::BasicBlock const *block = *toVisit.begin();
+			toVisit.pop_front();
+
+			if (visited.count(block))
+				continue;
+
+			if (std::optional<Stack> exitLayout = getExitLayoutOrStageDependencies(*block, visited, toVisit))
+			{
+				visited.emplace(block);
+				// We can skip the visit, if we have seen this precise exit layout already last time.
+				// Note: if the entire graph is revisited in the backwards jump check below, doing
+				//       this seems to break things; not sure why.
+				// Note: since I don't quite understand why doing this can break things, I comment
+				//       it out for now, since not aborting in those cases should always be safe.
+				// if (auto* previousInfo = util::valueOrNullptr(m_layout.blockInfos, block))
+				//	if (previousInfo->exitLayout == *exitLayout)
+				//		continue;
+				auto& info = m_layout.blockInfos[block];
+				info.exitLayout = *exitLayout;
+				info.entryLayout = propagateStackThroughBlock(info.exitLayout, *block);
+
+				for (auto entry: block->entries)
+					toVisit.emplace_back(entry);
+			}
+			else
+				continue;
+		}
+
+		// Determine which backwards jumps still require fixing and stage revisits of appropriate nodes.
+		for (auto [block, target]: backwardsJumps)
+			if (ranges::any_of(
+				m_layout.blockInfos[target].entryLayout,
+				[exitLayout = m_layout.blockInfos[block].exitLayout](StackSlot const& _slot) {
+					return !util::findOffset(exitLayout, _slot);
+				}
+			))
+			{
+				// This block jumps backwards, but does not provide all slots required by the jump target on exit.
+				// Therefore we need to visit the subgraph between ``target`` and ``block`` again.
+				// In particular we can visit backwards starting from ``block`` and mark all entries to-be-visited-
+				// again until we hit ``target``.
+				toVisit.emplace_front(block);
+				// Since we are likely to change the entry layout of ``target``, we also visit its entries again.
```

Comment:
Why does this not have to be done recursively?

Reply:
It *can* be done recursively (i.e. I can do ``visited.clear()`` and just revisit everything here), but going beyond the entries of ``target`` doesn't have much of an effect anymore in practice.
The *set* of slots will match, it's just the order that can change, I should probably clarify that and change "change" to "permute" or sth. It's also fine not to visit those entries here, but then I likely have to permute that stack before actually jumping. Basically this visit moves that permutation out of the loop condition one block up, which makes sense, since the loop condition probably will be executed more often than whatever jumped to it.

Reply:
If I want to avoid this kind of permuting at all cost I need to revisit the entire graph, but I tried and the benefit is marginal, if existant.

Reply:
I just thought doing it recursively is a requirement for correctness.

Reply:
Yeah, I should clarify that it's not.